File: frontend/src/api/bookingService.js
import api from './axiosInstance'

/**
 * Creates a new booking request.
 * @param {Object} payload { resourceId, startDateTime, endDateTime, purpose, attendees }
 */
export const createBooking = async (payload) => {
  const res = await api.post('/bookings', payload)
  return res.data
}

/**
 * Fetches the current user's bookings.
 */
export const getMyBookings = async () => {
  const res = await api.get('/bookings/my')
  return res.data
}

/**
 * Fetches details of a specific booking.
 */
export const getBookingById = async (id) => {
  const res = await api.get(`/bookings/${id}`)
  return res.data
}

/**
 * Cancels a specific booking.
 */
export const cancelBooking = async (id, data) => {
  const response = await api.patch(`/bookings/${id}/cancel`, data)
  return response.data
}

/** Admin Endpoints */
export const getAllBookings = async (params) => {
  const response = await api.get('/bookings', { params })
  return response.data
}

export const updateBookingStatus = async (id, data) => {
  const response = await api.put(`/bookings/${id}/status`, data)
  return response.data
}

export const getConflictingBookings = async (id) => {
  const response = await api.get(`/bookings/${id}/conflicts`)
  return response.data
}

/**
 * Fetch bookings for a specific resource and date.
 * @param {string|number} resourceId
 * @param {string} date - ISO date (yyyy-MM-dd) or Date object
 */
export const getBookingsForResource = async (resourceId, date) => {
  const params = {}
  if (date) params.date = typeof date === 'string' ? date : new Date(date).toISOString().slice(0, 10)
  const response = await api.get(`/bookings/resource/${resourceId}`, { params })
  return response.data
}

File: backend/src/main/java/com/smartcampus/incident/service/BookingService.java
package com.smartcampus.incident.service;

import com.smartcampus.incident.dto.booking.CancelBookingRequest;
import com.smartcampus.incident.dto.booking.CreateBookingRequest;
import com.smartcampus.incident.dto.booking.BookingStatusRequest;
import com.smartcampus.incident.dto.booking.BookingResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import java.time.LocalDate;
import java.util.List;

public interface BookingService {
    BookingResponse createBooking(CreateBookingRequest request);

    List<BookingResponse> getMyBookings();

    Page<BookingResponse> getAllBookings(String status, Long resourceId, Long userId, LocalDate date,
            Pageable pageable);

    List<BookingResponse> getBookingsForResource(Long resourceId, LocalDate date);

    BookingResponse getBookingById(Long id);

    void updateBookingStatus(Long id, BookingStatusRequest request);

    void cancelBooking(Long id, CancelBookingRequest request);
    BookingResponse getBookingByToken(String token);
    List<BookingResponse> getConflictingBookings(Long id);
}
